src/batch/src/executor2/generic_exchange.rs

```diff
@@ -128,60 +136,76 @@ impl<CS: 'static + CreateSource> Executor2 for GenericExchangeExecutor2<CS> {
 
 impl<CS: 'static + CreateSource> GenericExchangeExecutor2<CS> {
     #[try_stream(boxed, ok = DataChunk, error = RwError)]
-    async fn do_execute(mut self: Box<Self>) {
-        loop {
-            if self.source_idx >= self.sources.len() {
-                break;
-            }
-            if self.current_source.is_none() {
-                let proto_source = &self.sources[self.source_idx];
-                let source =
-                    CS::create_source(self.env.clone(), proto_source, self.task_id.clone()).await?;
-                self.current_source = Some(source);
+    async fn do_execute(self: Box<Self>) {
+        let mut sources: Vec<Box<dyn ExchangeSource>> = vec![];
+
+        for prost_source in &self.sources {
+            let source =
+                CS::create_source(self.env.clone(), prost_source, self.task_id.clone()).await?;
+            sources.push(source);
+        }
+
+        let mut handles = sources
+            .into_iter()
+            .map(data_chunk_stream)
+            .map(|stream| stream.into_future())
+            .collect_vec();
+
+        while !handles.is_empty() {
+            let ((result, from), _id, remainings) = select_all(handles).await;
```

Comment:
May use `futures::stream::select_all` so that you don't need to push the `from` into the `handles` manually.
Reference: https://github.com/singularity-data/risingwave-legacy/pull/736/files

Reply:
:joy: it is not open for me

Reply:
Sorry! May check this out, https://github.com/singularity-data/risingwave/blob/main/src/stream/src/executor/union.rs#L95

Reply:
It's just a different API that merges multiple streams together into one stream. And the caller can poll all the streams at the same time just like polling a single stream.